adr/0011-python-application-deployment.md

```diff
@@ -0,0 +1,77 @@
+how these cases are solved, the formula is built and maintained by those building and maintaining the systems these
+projects run on in production.
+
+Python projects require a Python runtime and Python packages can be installed system-wide or locally using a tool called
+'virtualenv', now officially part of Python3. Dependency and runtime management has become complicated in the move from
+Python2 to Python3 and the introduction of new language features and dependency management tools; the OS distributions 
+used in production use a fixed version of the Python runtime, often with features missing or mutually exclusive with 
+other features (Ubuntu + Python3 + pip + venv for example). Attempts to accommodate the latest runtimes has introduced
+instability and complexity in the rest of the system.
+
+Containerisation has made describing and building Python applications much easier for a project's author and the line 
+between traditional code development and code deployment is less clear. Container descriptions often live within the 
+code repository instead of in a separate one. Containers serve a single purpose, running a single process, but are 
+composable so that a cluster of containers may be described: one for the application, another for a webserver, another
+for a database and so on.
+
+Deploying a container that has the project within perfectly installed and configured is very similar to deploying a fat
+application binary: a single discrete artifact with all dependencies accounted for. Deployment consists of dropping the
+artifact into the runtime and reloading a service.
+
+Because container descriptions live alongside the codebase and are used during development they suffer the same 
+problems as the author-preferred utility scripts/glue code. There are any number of ways to build a container, does 
+installation/testing/linting/etc happen within the container or without? Are the build commands invoked in the same way 
+across projects? If we insist on containers, then is it the project author's responsibility to now maintain a suite of 
+them? Are they the best party to attempt it.
+
+## Decision
+
+Deploy all Python projects within containers.
+
+Testing happens within project containers.
+
+Standardise building of containers.
+
+## Consequences
+
+Testing becomes less hazardous as fewer variables are in play.
+
+Deployment becomes simpler and much of a project's formula can be removed.
+
+Deployment becomes uniform and new projects can benefit from the tooling built for other Python projects.
+
+Adhoc and idiomatic utility scripts are replaced with a single script that encapsulates build and test commands.
+
+Project authors will have a reproducible environment much closer to production and much less unwieldy than a 
+Vagrant virtual machine.
+
+Project releases have a reference environment that non-eLife parties can use (that doesn't require heavy dependencies like Builder and Vagrant).
```

Comment:
This arguably is not fitting with the rest, only the `libero` Github organization targets reuse but doesn't follow this `tech-team`'s repository rules

Reply:
I know you make that distinction. I'll drop the second one but I think the first has real merit. It isn't a goal but more of a beneficial side effect.